src/main/java/org/opensearch/knn/index/codec/KNN80Codec/KNN80DocValuesConsumer.java

```diff
@@ -184,6 +188,20 @@ private void createKNNIndexFromScratch(FieldInfo fieldInfo, KNNCodecUtil.Pair pa
 
         // Used to determine how many threads to use when indexing
         parameters.put(KNNConstants.INDEX_THREAD_QTY, KNNSettings.state().getSettingValue(KNNSettings.KNN_ALGO_PARAM_INDEX_THREAD_QTY));
+        Object name = parameters.get(NAME);
+        if (name != null && name.equals(METHOD_NSG)) {
+            if (parameters.containsKey(PARAMETERS)) {
+                parameters.remove(PARAMETERS);
+            }
+            /** TODO:
+             * when numIds is too small, NSG graph would core/throw exception
+             * because There are too much invalid entries in the knn graph.
```

Comment:
Interesting. Is there any documentation on this?

Reply:
When NSG graph n is too small , it will receive `SIGFPE` signal as following picture shows:

![img_v2_70b65047-33a9-4772-9f20-d51845da4e8g](https://github.com/opensearch-project/k-NN/assets/12760367/2093b1c1-cc0a-4f90-85b2-ac1fc1bcb3ec)

this because default `build_type=1` build with NNDescent, when call  `NNDescent#gen_random` and
`addr[i] = rng() % (N - size);` N - size would be zero when N is small in the code:
https://github.com/facebookresearch/faiss/blob/3219e3d12e6fc36dfdfe17d4cf238ef70bf89568/faiss/impl/NNDescent.cpp#L90-L93

Even if we use `build_type = 0` with brute force search
https://github.com/facebookresearch/faiss/blob/3219e3d12e6fc36dfdfe17d4cf238ef70bf89568/faiss/IndexNSG.cpp#L157-L163

it will throw exception in 
`check_knn_graph(knng.data(), n, GK);` that is 
```
    FAISS_THROW_IF_NOT_MSG(
            total_count < n / 10,
            "There are too much invalid entries in the knn graph. "
            "It may be an invalid knn graph.");
```
https://github.com/facebookresearch/faiss/blob/3219e3d12e6fc36dfdfe17d4cf238ef70bf89568/faiss/IndexNSG.cpp#L238


It is because NSG needs some entry points for search, and in small dataset and positive GK parameter, many invalid graph entries generated. 

simple tests in previous commits for  testing this scenario: https://github.com/opensearch-project/k-NN/blob/639f90c3273bf6f26731fe67b99a0f6a9dc251a5/src/test/java/org/opensearch/knn/index/FaissIT.java#L195-L202

so i try to solve this scenario, i use a FLAT map for small dataset for avoid throw exception when index only 1~256 documents.

Reply:
Sounds good. Im still not sure why in NNDescent the range is :`N - size` for the random index and not just N. Isnt the purpose to just select random neighbors in the dataset for the neighborhood? Why would it need to be `- size`? Is this possibly a bug?

Additionally, on the brute force approach, it seems wrong to check invalid entries based on GK. If GK is set very high for small data, the graph is not invalid. Brute force approach did create the correct graph. 

I might be missing something, but it seems like this could be made more robust on *faiss* side?

Reply:
@luyuncheng Can you look at this as well?

Reply:
When i use  brute force approach and remove the check it would enter into endless loop in https://github.com/facebookresearch/faiss/blob/3219e3d12e6fc36dfdfe17d4cf238ef70bf89568/faiss/impl/NSG.cpp#L296-L305  which is try to find enough entries
```
    while (num_ids < pool_size) {
        int id = gen.rand_int(ntotal);
        if (vt.get(id)) {
            continue;
        }

        init_ids[num_ids] = id;
        num_ids++;
        vt.set(id);
    }
```
![image](https://github.com/opensearch-project/k-NN/assets/12760367/3f38cfb6-5e04-4eaa-9242-7e14fd8f62e3)

----

@jmazanec15  FROM my idea: 
NSG based on `NNDescent`, there is an `assumed that K << N` so there is must ensure enough initial neighbor in the graph. when we using tests with N=1 or N << K, it can not build a initial graph for finding enough entries. 

![image](https://github.com/opensearch-project/k-NN/assets/12760367/6409c0e9-ae86-4f90-81e3-e02a66d3c468)